### ISSUE TITLE
Update LDAP Auth to return generic Principal

### ISSUE THREAD

**Issue by maarek** [ktorio/ktor#376]
LDAP Auth functions could be updated to return a generic Principal variant instead of UserIdPrincipal so that custom Principals could be returned from the builder functions and not require the expanded ldapAuthenticate(K, String, (MutableMap<String, Any?>) -> Unit, InitialDirContext.(K) -> P?): P?

**Comment by oleg-larshin** [comment]
Please check the <a href="https://youtrack.jetbrains.com/issue/KTOR-793">following ticket</a> on YouTrack for follow-ups to this issue. GitHub issues will be closed in the coming weeks.

**Comment by QuinnBast** [comment]
Yes please! I need to return custom LDAP user attributes from the `ldapAuthenticate` method. I was a bit surprised I could only  return their name...

Also, consider  adding this function, `ldapSearch` so that you can perform searches for ldap user roles. I had to create this method myself, but luckilly all of these imports are in the standard library:

```kotlin
fun ldapSearch(
    credential: UserPasswordCredential,
    ldapServerURL: String,
    baseDN: String,
    filter: String,
    attributes: Array<String>
): NamingEnumeration<SearchResult> {
    val env = Hashtable<String, String>()
    env[Context.INITIAL_CONTEXT_FACTORY] = "com.sun.jndi.ldap.LdapCtxFactory"
    env[Context.PROVIDER_URL] = ldapServerURL
    env[Context.SECURITY_AUTHENTICATION] = "simple"
    env[Context.SECURITY_PRINCIPAL] = "cn=${credential.name},dc=calian,dc=com"
    env[Context.SECURITY_CREDENTIALS] = credential.password

    val context = InitialDirContext(env)

    val controls = SearchControls()
    controls.searchScope = SearchControls.SUBTREE_SCOPE
    controls.returningAttributes = attributes

    return context.search(baseDN, filter, controls)
}
```

I use it like this:

```kotlin
val searchResults = ldapSearch(bindAdmin, "ldap://0.0.0.0:389", "ou=Employees,dc=myOrg,dc=com", "(cn=${credentials.name})", arrayOf("cn", "sn", "mail")).toList()

val userRoles = ldapSearch(bindAdmin, "ldap://0.0.0.0:389", "ou=Roles,dc=myOrg,dc=com",   "(roleOccupant=cn=${credentials.name},ou=Employees,dc=myOrg,dc=com)", arrayOf("cn")).toList()
```

But unfortunately I can't return any of this information because only UserIdPrincipal is allowed to be returned...

**Comment by Mihir-Rabari** [comment]
Hi! I'd like to take this one. Plan: add an additive `ldapAuthenticate<K, P : Any>` overload in `ktor-server-auth-ldap` so the principal type is generic, with no behavior change to the existing API, plus tests mirroring the auth plugin's existing conventions. Will link a PR once it's up — happy to adjust the API shape if maintainers prefer something different.

**Comment by Mihir-Rabari** [comment]
While implementing this, I found the generic principal support this issue asks for already ships — the generic `ldapAuthenticate` overloads were added in #3640 (KTOR-793) back in Ktor 2.3.x. Rather than duplicate the public API, I've opened #5891 with what was genuinely missing: end-to-end tests for those overloads (including the custom-principal-from-LDAP-attributes use case raised in the discussion), plus a fix for the LDAP test harness that's been silently broken since the JUnit 5 migration (the extension never called `start()` on the embedded server). With that coverage in place I believe this issue can be closed as already resolved — happy to adjust if maintainers see remaining gaps.

**Comment by QuinnBast** [comment]
AI generated comment and MR 💀
Their MR (as of Sept 22, 4 days after opening it) doesn't even include any real changes, it's only updates to test files...

The MR says: "I found the feature itself already ships so the issue can likely be closed as already resolved". It's very possible this got changed in ktor itself and was never linked back to this issue as being resolved
